subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java

```diff
@@ -214,13 +215,19 @@ public void execute(ExecutionListener executionListener) {
                 ParsedCommandLine parsedCommandLine = parser.parse(args);
 
                 buildLayoutConverter.convert(parsedCommandLine, buildLayout);
-                loggingConfigurationConverter.convert(parsedCommandLine, loggingConfiguration);
+
 
                 Map<String, String> properties = new HashMap<String, String>();
                 // Read *.properties files
                 layoutToPropertiesConverter.convert(buildLayout, properties);
                 // Read -D command line flags
                 systemPropertiesCommandLineConverter.convert(parsedCommandLine, properties);
+
+                // Convert properties for logging  object
+                PropertiesToLogLevelConfigurationConverter propertiesToLogLevelConfigurationConverter = new PropertiesToLogLevelConfigurationConverter();
```

Comment:
We should not keep adding more ad hoc implementations for build options that are available both as a command-line and as a property. This leads to all kinds of inconsistent behaviour.

Instead, we should introduce a type that represents a build option, and use this definition to consistently apply handling from the command-line, system properties, environment variables, `gradle.properties` and so on.

I would start growing `GradleProperties` into a collection of `GradleBuildOption` objects instead of a collection of strings. Each build option object would specify:

- the option type: which can just be a boolean or a string for now.
- the command-line option to specify the option, and a description.
- the Gradle property to specify the option.
- a method that takes a `StartParameter` and the value of the specified type, whose responsibility is to attach the option's value on the `StartParameter`, eg by calling the appropriate setter.

No every option would have a specific command-line option.

From these definitions, we would configure the command-line parser. For boolean typed options, we would add a `--no-<foo>` flag for the options, as well as a `--<foo>` flag. For other options, we would add a `--<foo>` flag that takes a value. We would do this without hard-coding anything - the code would simply iterate over the set of option definitions and define the command-line flags based on the data provided by the option definition.

From these definitions, we would also look for properties in various locations, applying the correct order of precedence.